tests/harness/utils.py

```diff
@@ -0,0 +1,349 @@
+            Ignored if an environment name is specified.
+        no_provision: Whether to skip environment provisioning (assumes
+            environment is already running).
+        no_teardown: Whether to skip environment teardown on exit. If the
+            environment is already running on entry, it will not be torn down.
+        no_deprovision: Whether to skip environment deprovisioning on exit.
+            If the environment is already provisioned on entry, it will not be
+            deprovisioned on exit.
+
+    Yields:
+        The active environment and a client connected with it.
+    """
+
+    # set env variables
+    os.environ[ENV_ZENML_DEBUG] = "true"
+    os.environ["ZENML_ANALYTICS_OPT_IN"] = "false"
+
+    # original working directory
+    orig_cwd = os.getcwd()
+
+    harness = TestHarness()
+
+    environment = harness.set_environment(
+        environment_name=environment_name,
+        deployment_name=deployment_name,
+        requirements_names=requirements_names,
+    )
+
+    if no_provision:
+        logging.info("Skipping environment provisioning")
+        with environment.deployment.connect() as client:
+            yield environment, client
+    else:
+        # Provision the environment (bring up the deployment, if local, and
+        # register the stacks according to the environment's configuration)
+        with environment.setup(
+            teardown=not no_teardown,
+            deprovision=not no_deprovision,
+        ) as client:
+            logging.info(
+                f"Test session is using environment '{environment.config.name}' "
+                f"running at '{client.zen_store.url}'."
+            )
+
+            yield environment, client
+
+    # change working directory back to base path
+    os.chdir(orig_cwd)
+
+
+@contextmanager
+def clean_repo_session(
+    tmp_path_factory: pytest.TempPathFactory,
+) -> Generator[Client, None, None]:
+    """Context manager to initialize and use a separate ZenML repository.
+
+    Args:
+        tmp_path_factory: A pytest fixture that provides a temporary directory.
+
+    Yields:
+        A ZenML client connected to the repository.
+    """
+
+    # original working directory
+    orig_cwd = os.getcwd()
+
+    # change the working directory to a fresh temp path
+    tmp_path = tmp_path_factory.mktemp("pytest-zenml-repo")
+    os.chdir(tmp_path)
+
+    client = Client()
+    orig_root = client.root
+
+    client.initialize(tmp_path)
+    client.activate_root(tmp_path)
+
+    logging.info(f"Tests are running in clean repository: '{tmp_path}'")
+
+    yield client
+
+    # remove all traces, and change working directory back to base path
+    os.chdir(orig_cwd)
+    client.activate_root(orig_root)
+    cleanup_folder(str(tmp_path))
+
+
+@contextmanager
+def clean_project_session(
+    tmp_path_factory: pytest.TempPathFactory,
+    clean_repo: bool = False,
+) -> Generator[Client, None, None]:
```

Comment:
Sorry to ask this here, how exactly is the scoping for different tests designed?

Environment > Client/Root > Project ?

Reply:
Different tests use different fixtures in `conftest.py` files that are located in various places in the test tree:

* the examples use the same Zen store, but a different repository for each test because they use different active stacks
* the CLI tests use the same Zen store, but different repositories with different projects for different tests, depending on what they need
* zen store tests use separate sql zen stores for each test
* unit-tests are still in the original state of using a separate clean client for each module